File: firebase-firestore/src/main/java/com/google/firebase/firestore/auth/FirebaseAuthCredentialsProvider.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.firebase.firestore.auth;

import android.annotation.SuppressLint;
import androidx.annotation.GuardedBy;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.FirebaseApiNotAvailableException;
import com.google.firebase.FirebaseApp;
import com.google.firebase.auth.GetTokenResult;
import com.google.firebase.auth.internal.IdTokenListener;
import com.google.firebase.auth.internal.InternalAuthProvider;
import com.google.firebase.firestore.util.Executors;
import com.google.firebase.firestore.util.Listener;
import com.google.firebase.firestore.util.Logger;
import com.google.firebase.inject.Deferred;
import com.google.firebase.inject.Provider;

/**
 * FirebaseAuthCredentialsProvider uses Firebase Auth via {@link FirebaseApp} to get an auth token.
 *
 * <p>NOTE: To simplify the implementation, it requires that you call {@link #setChangeListener} no
 * more than once and don't call {@link #getToken} after calling {@link #removeChangeListener}.
 *
 * <p>This class must be implemented to be thread-safe since getToken() and
 * set/removeChangeListener() are called from the Firestore worker thread, but the getToken() Task
 * callbacks and user change notifications will be executed on arbitrary different threads.
 */
public final class FirebaseAuthCredentialsProvider extends CredentialsProvider<User> {

  private static final String LOG_TAG = "FirebaseAuthCredentialsProvider";

  /**
   * The listener registered with FirebaseApp; used to stop receiving auth changes once
   * changeListener is removed.
   */
  private final IdTokenListener idTokenListener = result -> onIdTokenChanged();

  /**
   * The {@link Provider} that gives access to the {@link InternalAuthProvider} instance; initially,
   * its {@link Provider#get} method returns {@code null}, but will be changed to a new {@link
   * Provider} once the "auth" module becomes available.
   */
  @Nullable
  @GuardedBy("this")
  private InternalAuthProvider internalAuthProvider;

  /** The listener to be notified of credential changes (sign-in / sign-out, token changes). */
  @Nullable
  @GuardedBy("this")
  private Listener<User> changeListener;

  /** Counter used to detect if the token changed while a getToken request was outstanding. */
  @GuardedBy("this")
  private int tokenCounter;

  @GuardedBy("this")
  private boolean forceRefresh;

  /** Creates a new FirebaseAuthCredentialsProvider. */
  @SuppressLint("ProviderAssignment") // TODO: Remove this @SuppressLint once b/181014061 is fixed.
  public FirebaseAuthCredentialsProvider(Deferred<InternalAuthProvider> deferredAuthProvider) {
    deferredAuthProvider.whenAvailable(
        provider -> {
          synchronized (this) {
            internalAuthProvider = provider.get();
            onIdTokenChanged();
            internalAuthProvider.addIdTokenListener(idTokenListener);
          }
        });
  }

  @Override
  public synchronized Task<String> getToken() {
    if (internalAuthProvider == null) {
      return Tasks.forException(new FirebaseApiNotAvailableException("auth is not available"));
    }

    Task<GetTokenResult> res = internalAuthProvider.getAccessToken(forceRefresh);
    forceRefresh = false;

    // Take note of the current value of the tokenCounter so that this method can fail (with a
    // FirebaseFirestoreException) if there is a token change while the request is outstanding.
    final int savedCounter = tokenCounter;
    return res.continueWithTask(
        Executors.DIRECT_EXECUTOR,
        task -> {
          synchronized (this) {
            // Cancel the request since the token changed while the request was outstanding so the
            // response is potentially for a previous user (which user, we can't be sure).
            if (savedCounter != tokenCounter) {
              Logger.debug(LOG_TAG, "getToken aborted due to token change");
              return getToken();
            }

            if (task.isSuccessful()) {
              return Tasks.forResult(task.getResult().getToken());
            } else {
              return Tasks.forException(task.getException());
            }
          }
        });
  }

  @Override
  public synchronized void invalidateToken() {
    forceRefresh = true;
  }

  @Override
  public synchronized void setChangeListener(@NonNull Listener<User> changeListener) {
    this.changeListener = changeListener;

    // Fire the initial event.
    changeListener.onValue(getUser());
  }

  @Override
  public synchronized void removeChangeListener() {
    changeListener = null;

    if (internalAuthProvider != null) {
      internalAuthProvider.removeIdTokenListener(idTokenListener);
    }
  }

  /** Invoked when the auth token changes. */
  private synchronized void onIdTokenChanged() {
    tokenCounter++;
    if (changeListener != null) {
      changeListener.onValue(getUser());
    }
  }

  /** Returns the current {@link User} as obtained from the given InternalAuthProvider. */
  private synchronized User getUser() {
    @Nullable String uid = (internalAuthProvider == null) ? null : internalAuthProvider.getUid();
    return uid != null ? new User(uid) : User.UNAUTHENTICATED;
  }
}

File: firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/auth/FirebaseAuthCredentialsProvider.kt
package com.google.firebase.dataconnect.auth

import com.google.firebase.auth.internal.IdTokenListener
import com.google.firebase.auth.internal.InternalAuthProvider
import com.google.firebase.dataconnect.util.Listener
import com.google.firebase.inject.Deferred
import com.google.firebase.inject.Provider
import java.util.concurrent.atomic.AtomicLong
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.getAndUpdate
import kotlinx.coroutines.tasks.await

public interface FirebaseAuthCredentialsProvider : CredentialsProvider<User>

/**
 * FirebaseAuthCredentialsProvider uses Firebase Auth via {@link FirebaseApp} to get an auth token.
 */
public fun FirebaseAuthCredentialsProvider(
  deferredProvider: Deferred<InternalAuthProvider>
): FirebaseAuthCredentialsProvider = FirebaseAuthCredentialsProviderImpl(deferredProvider)

internal class FirebaseAuthCredentialsProviderImpl(
  deferredProvider: Deferred<InternalAuthProvider>
) : FirebaseAuthCredentialsProvider {

  @OptIn(ExperimentalCoroutinesApi::class, DelicateCoroutinesApi::class)
  private val singleThreadDispatcher =
    newSingleThreadContext("FirebaseAuthCredentialsProviderThread")

  /**
   * The {@link Provider} that gives access to the {@link InternalAuthProvider} instance; initially,
   * its {@link Provider#get} method returns {@code null}, but will be changed to a new {@link
   * Provider} once the "auth" module becomes available.
   */
  private lateinit var internalAuthProvider: InternalAuthProvider

  /**
   * The listener registered with FirebaseApp; used to stop receiving auth changes once
   * changeListener is removed.
   */
  private val idTokenListener = IdTokenListener { onIdTokenChanged() }

  /** The listener to be notified of credential changes (sign-in / sign-out, token changes). */
  private var changeListener: Listener<User>? = null

  private var forceRefresh = MutableStateFlow(false)

  /** Id used to detect if the token changed while a getToken request was outstanding. */
  private val tokenId = AtomicLong(0)

  /** Creates a new FirebaseAuthCredentialsProvider. */
  init {
    deferredProvider.whenAvailable { provider: Provider<InternalAuthProvider> ->
      runBlocking {
        withContext(singleThreadDispatcher) {
          internalAuthProvider = provider.get()
          internalAuthProvider.addIdTokenListener(idTokenListener)
        }
        onIdTokenChanged()
      }
    }
  }

  override suspend fun getToken(): String? {
    var isFinished = false
    var result: String? = null
    withContext(singleThreadDispatcher) {
      if (!this@FirebaseAuthCredentialsProviderImpl::internalAuthProvider.isInitialized) {
        // Auth is not available
        isFinished = true
        return@withContext
      }

      val currentForceRefresh = forceRefresh.getAndUpdate { false }
      val res = internalAuthProvider.getAccessToken(currentForceRefresh)

      val currentTokenId = tokenId.get()
      result =
        try {
          res.await().token
        } catch (_: Exception) {
          null
        }

      if (currentTokenId == tokenId.get()) {
        isFinished = true
      }
    }
    return if (isFinished) result else getToken()
  }

  override suspend fun invalidateToken() {
    forceRefresh.value = true
  }

  override suspend fun removeChangeListener() {
    withContext(singleThreadDispatcher) {
      changeListener = null
      if (this@FirebaseAuthCredentialsProviderImpl::internalAuthProvider.isInitialized) {
        internalAuthProvider.removeIdTokenListener(idTokenListener)
      }
    }
  }

  override suspend fun setChangeListener(changeListener: Listener<User>) {
    withContext(singleThreadDispatcher) {
      this@FirebaseAuthCredentialsProviderImpl.changeListener = changeListener
    }
    // Fire the initial event.
    changeListener.onValue(getUser())
  }

  /** Invoked when the auth token changes. */
  private fun onIdTokenChanged() {
    runBlocking {
      tokenId.incrementAndGet()
      if (changeListener != null) {
        changeListener!!.onValue(getUser())
      }
    }
  }

  /**
   * Returns the current {@link User} as obtained from the given InternalAuthProvider. This function
   * can only be called in functions which runs within singleThreadDispatcher.
   */
  private suspend fun getUser(): User {
    return withContext(singleThreadDispatcher) {
      val uid =
        if (!this@FirebaseAuthCredentialsProviderImpl::internalAuthProvider.isInitialized) {
          null
        } else {
          internalAuthProvider.uid
        }
      return@withContext if (uid != null) User(uid) else User.UNAUTHENTICATED
    }
  }
}
